File: Unity/TSProj/src/mods/ModTheSpire/Scripts/Unit/Monster/AbstractMonster.ts
import { Mathf } from "../../../../../Core/Module/Math/Mathf";
import TWEEN from "../../../../../ThirdParty/@tweenjs/tween";
import DamageInfo from "../../DataDefine/DamageInfo";
import DungeonManager from "../../DungeonManager";
import { OnPreBattleMessage } from "../../Events/OnPreBattleMessage";
import AbstractCreature from "../AbstractCreature";
import { Intent } from "./Intent";
import { MonsterInfo } from "./MonsterInfo";

export abstract class AbstractMonster extends AbstractCreature
{
    public MoveName:string;
    public Gold : number;
    public MoveIndex:number;
    public DeathTime:number;
    public Intent:Intent;
    //行动历史
    private MoveHistory = new Array<number>();
    //伤害列表,不同难度填充的伤害内容不一样
    protected DamageInfo = new Array<DamageInfo>()
    constructor(public Info:MonsterInfo) {
        super();
        this.DisplayObject = this.Info.Loader.Create();
        this.Initialize();
        this.RollMove();
    }
    //随机敌人的行为
    public RollMove(){
        this.GetMove(Mathf.Floor(Mathf.RandomRange(0,101,DungeonManager.Inst.CurrentDungeon.AIRng)));
    }
    //获取敌人的最后两次行为是否和传入的行为Index匹配
    public LastTwoMoves(num:number) : boolean{
        let length = this.MoveHistory.length;
        if(length < 2)
            return false;
        return this.MoveHistory[length - 2] == num && this.MoveHistory[length - 1] == num;
    }
    //获取敌人的最后一次行为是否和传入的行为Index匹配
    public LastMove(num:number) : boolean{
        let length = this.MoveHistory.length;
        if(length < 1)
            return false;
        return this.MoveHistory[length - 1] == num;
    }
    //初始化.继承此类的类型不应该使用构造函数.而应该选择使用Initialize作为替代;
    abstract Initialize();
    //敌人的行为逻辑代码
    abstract TakeTurn();
    //获取敌人的下一次行动
    abstract GetMove(num:number);
    //设置敌人的下一次行动
    public SetMove(moveIndex:number,intent:Intent,options?:{damage?:DamageInfo,multiplier?:number,moveName?:string}){
        this.MoveName = options.moveName;
        if(moveIndex != -1)
            this.MoveHistory.push(moveIndex);
        this.MoveIndex = moveIndex;
        this.Intent = intent;
    }
    //敌人死亡回调
    Die(){};
    public Escape(){
        this.HealthComponent.Visible = false;
        this.IsEscaping = true;
        this.DisplayObject.FlipX = true;
        new TWEEN.Tween({x:this.X}).to({x:this.X + 400}).easing(TWEEN.Easing.Linear.None).duration(3000).start().onComplete(()=>{
            this.IsEscaped = true;
            if(DungeonManager.Inst.CurrentRoom.Monsters.AreMonstersBasicallyDead())
                DungeonManager.Inst.CurrentRoom.End();
        });
    }
    PreBattleAction(msg:OnPreBattleMessage){};
}

File: Unity/Assets/Scripts/ModTheSpire/Unit/Monster/AbstractMonster.cs
using System.Collections.Generic;
using DG.Tweening;
using ModTheSpire.DataDefine;
using ModTheSpire.Dungeon;
using ModTheSpire.Event.Ids;
using UnityEngine;

namespace ModTheSpire.Unit.Monster
{
    public abstract class AbstractMonster : AbstractCreature
    {
        public MonsterInfo Info;
        public string MoveName;
        public int Gold;
        public int MoveIndex;
        public int DeathTime;

        public Intent Intent;

        //行动历史
        private List<int> MoveHistory = new();

        //伤害列表,不同难度填充的伤害内容不一样
        protected List<DamageInfo> DamageInfo = new();

        public AbstractMonster(MonsterInfo info) : base()
        {
            Info = info;
            this.DisplayObject = this.Info.Loader.Create();
            this.Initialize();
            this.RollMove();
        }

        //随机敌人的行为
        public virtual void RollMove()
        {
            this.GetMove(Mathf.FloorToInt(DungeonManager.Inst.CurrentDungeon.AIRng.Next(0, 101)));
        }

        //获取敌人的最后两次行为是否和传入的行为Index匹配
        public virtual bool LastTwoMoves(int num)
        {
            var length = this.MoveHistory.Count;
            if (length < 2)
                return false;
            return this.MoveHistory[length - 2] == num && this.MoveHistory[length - 1] == num;
        }

        //获取敌人的最后一次行为是否和传入的行为Index匹配
        public bool LastMove(int num)
        {
            var length = this.MoveHistory.Count;
            if (length < 1)
                return false;
            return this.MoveHistory[length - 1] == num;
        }

        //初始化.继承此类的类型不应该使用构造函数.而应该选择使用Initialize作为替代;
        protected abstract void Initialize();

        //敌人的行为逻辑代码
        protected abstract void TakeTurn();

        //获取敌人的下一次行动
        protected abstract void GetMove(int num);

        //设置敌人的下一次行动
        public void SetMove(int moveIndex, Intent intent, DamageInfo damage = null, int multiplier = 0, string moveName = null)
        {
            MoveName = moveName;
            if (moveIndex != -1)
                this.MoveHistory.Add(moveIndex);
            this.MoveIndex = moveIndex;
            this.Intent = intent;
        }

        //敌人死亡回调
        protected virtual void Die()
        {
        }

        public void Escape()
        {
            this.DisplayObject.HealthComponent.Visible = false;
            this.IsEscaping = true;
            this.DisplayObject.FlipX = true;
            DOTween.To(() => this.DisplayObject.X, t1 => this.DisplayObject.X = t1, this.DisplayObject.X + 400, 3).SetEase(Ease.Linear).OnComplete(
                () =>
                {
                    this.IsEscaped = true;
                    if (DungeonManager.Inst.CurrentRoom.Monsters.AreMonstersBasicallyDead())
                        DungeonManager.Inst.CurrentRoom.End();
                });
        }

        protected virtual void PreBattleAction(OnPreBattleMessage msg)
        {
        }
    }
}